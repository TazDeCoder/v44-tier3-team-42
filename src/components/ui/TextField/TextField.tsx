'use client';

/* eslint-disable react/jsx-props-no-spreading */
import type {
  ForwardRefExoticComponent,
  HTMLInputTypeAttribute,
  InputHTMLAttributes,
  RefAttributes,
} from 'react';
import { forwardRef } from 'react';
import { twMerge } from 'tailwind-merge';

const styles = {
  root: '',
  label: 'text-sm leading-tight tracking-wide text-subtleLight',
  input: 'text-base leading-normal tracking-normal text-subtleDark',
} as const;

export type TextFieldProps = InputHTMLAttributes<HTMLInputElement> & {
  /**
   * Label text associated to component
   */
  label: string;
  /**
   * Trailing icon for component
   */
  trailIcon?: React.ReactNode;
  type?: Extract<HTMLInputTypeAttribute, 'text' | 'password'>;
};

export const TextField: ForwardRefExoticComponent<
  TextFieldProps & RefAttributes<HTMLInputElement>
> = forwardRef(
  ({ id, label, type = 'text', trailIcon: Icon, className, ...props }, ref) => (
    <label className={styles.root} htmlFor={id}>
      <p className={styles.label}>{label}</p>
      <div className="flex items-center h-12 p-3 mt-2 border rounded-lg gap-x-4 border-subtleLight">
        <input
          ref={ref}
          id={id}
          type={type}
          className={twMerge(styles.input, className)}
          {...props}
        />
        {Icon || null}
      </div>
    </label>
  ),
);

if (process.env.NODE_ENV !== 'production') {
  TextField.displayName = 'TextField';
}
