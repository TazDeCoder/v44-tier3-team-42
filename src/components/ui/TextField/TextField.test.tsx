import { render, screen } from 'test-utils';
import userEvent from '@testing-library/user-event';
import { Show } from 'react-iconly';

import { TextField } from './TextField';

describe('<TextField />', () => {
  describe('Snapshots', () => {
    it('should match snapshot', () => {
      const { asFragment } = render(<TextField label="Foo" />);
      expect(asFragment()).toMatchSnapshot();
    });
    it('should match snapshot in default text-field with trailIcon prop', () => {
      const { asFragment } = render(
        <TextField label="Foo" trailIcon={<Show />} />,
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe('User Interactions', () => {
    it('should update value based on user input', async () => {
      render(<TextField label="Foo" />);
      const inputEl = screen.getByRole('textbox');
      await userEvent.click(inputEl);
      await userEvent.type(inputEl, 'example');
      expect(inputEl).toHaveValue('example');
    });
  });
}
)
