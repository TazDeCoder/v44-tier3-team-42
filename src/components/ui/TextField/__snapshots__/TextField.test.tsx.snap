// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> Snapshots should match snapshot 1`] = `
<DocumentFragment>
  <label
    class=""
  >
    <p
      class="text-sm leading-tight tracking-wide text-subtleLight"
    >
      Foo
    </p>
    <div
      class="flex items-center h-12 p-3 mt-2 border rounded-lg gap-x-4 border-subtleLight"
    >
      <input
        class="text-base leading-normal tracking-normal text-subtleDark"
        type="text"
      />
    </div>
  </label>
</DocumentFragment>
`;

exports[`<TextField /> Snapshots should match snapshot in default text-field with trailIcon prop 1`] = `
<DocumentFragment>
  <label
    class=""
  >
    <p
      class="text-sm leading-tight tracking-wide text-subtleLight"
    >
      Foo
    </p>
    <div
      class="flex items-center h-12 p-3 mt-2 border rounded-lg gap-x-4 border-subtleLight"
    >
      <input
        class="text-base leading-normal tracking-normal text-subtleDark"
        type="text"
      />
      <svg
        height="24px"
        role="presentation"
        viewBox="0 0 24 24"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          transform="translate(2 4)"
        >
          <path
            d="M6.323,3.162A3.162,3.162,0,1,1,3.161,0,3.162,3.162,0,0,1,6.323,3.162Z"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-miterlimit="10"
            stroke-width="1.5px"
            transform="translate(6.839 4.891)"
          />
          <path
            d="M9.248,14.6c3.808,0,7.291-2.738,9.252-7.3C16.539,2.738,13.056,0,9.248,0h0C5.444,0,1.961,2.738,0,7.3c1.961,4.564,5.444,7.3,9.252,7.3Z"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-miterlimit="10"
            stroke-width="1.5px"
            transform="translate(0.75 0.751)"
          />
        </g>
      </svg>
    </div>
  </label>
</DocumentFragment>
`;
